pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,287 @@
+			clientID:     "clientidd1",
+			errorMessage: nil,
+			pubsub:       false,
+			handler:      handler{false},
+		},
+		{
+			desc:         "Unsubscribe from an already unsubscribed topic with a subtopic with an ID",
+			topic:        fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID:     "clientid1",
+			errorMessage: mqtt.ErrNotSubscribed,
+			pubsub:       false,
+			handler:      handler{false},
+		},
+		{
+			desc:         "Subscribe to an empty topic with an ID",
+			topic:        "",
+			clientID:     "clientid1",
+			errorMessage: mqtt.ErrEmptyTopic,
+			pubsub:       true,
+			handler:      handler{false},
+		},
+		{
+			desc:         "Unsubscribe from an empty topic with an ID",
+			topic:        "",
+			clientID:     "clientid1",
+			errorMessage: mqtt.ErrEmptyTopic,
+			pubsub:       false,
+			handler:      handler{false},
+		},
+		{
+			desc:         "Subscribe to a topic with empty id",
+			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID:     "",
+			errorMessage: mqtt.ErrEmptyID,
+			pubsub:       true,
+			handler:      handler{false},
+		},
+		{
+			desc:         "Unsubscribe from a topic with empty id",
+			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID:     "",
+			errorMessage: mqtt.ErrEmptyID,
+			pubsub:       false,
+			handler:      handler{false},
+		},
+		{
+			desc:         "Subscribe to another topic with an ID",
+			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic+"1"),
+			clientID:     "clientid3",
+			errorMessage: nil,
+			pubsub:       true,
+			handler:      handler{true},
+		},
+		{
+			desc:         "Subscribe to another already subscribed topic with an ID with Unsubscribe failing",
+			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic+"1"),
+			clientID:     "clientid3",
+			errorMessage: mqtt.ErrFailed,
+			pubsub:       true,
+			handler:      handler{true},
+		},
+		{
+			desc:         "Subscribe to a new topic with an ID",
+			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic+"2"),
+			clientID:     "clientid4",
+			errorMessage: nil,
+			pubsub:       true,
+			handler:      handler{true},
+		},
+		{
+			desc:         "Unsubscribe from a topic with an ID with failing handler",
+			topic:        fmt.Sprintf("%s.%s", chansPrefix, topic+"2"),
+			clientID:     "clientid4",
+			errorMessage: mqtt.ErrFailed,
+			pubsub:       false,
+			handler:      handler{true},
+		},
+	}
+
+	for _, pc := range cases {
+		if pc.pubsub == true {
+			err := pubsub.Subscribe(pc.clientID, pc.topic, pc.handler)
+			if pc.errorMessage == nil {
+				require.Nil(t, err, fmt.Sprintf("%s got unexpected error: %s", pc.desc, err))
+			} else {
+				assert.Equal(t, err, pc.errorMessage)
+			}
+		} else {
+			err := pubsub.Unsubscribe(pc.clientID, pc.topic)
+			if pc.errorMessage == nil {
+				require.Nil(t, err, fmt.Sprintf("%s got unexpected error: %s", pc.desc, err))
+			} else {
+				assert.Equal(t, err, pc.errorMessage)
+			}
+		}
```

Comment:
This looks too complex, try to simplify.

Reply:
This is the way it's implemented for nats and rabbitmq also.
I'll simplify this, and also make the change in those two files as well.

Reply:
I have made some changes, one commented out, and other not commented out.
Please tell if either of them is better than the current version

Reply:
@AryanGodara `TestSubscriber` mixes testing `Subscribe` and `Unsubscribe` methods. TBH, I don't mind that because it's a part of the same functionality (even inseparable) but it's a little cleaner to have them separated. The complexity of executing test cases is an indication that we should split them. For `Subscribe` use the same principle as for the `Publish` test - use the plain client to publish messages you expect on the Subscriber side. For `Unsubscribe` - you obviously need to call `Subscribe` (you can do that before iterating through the cases, though), there is nothing we can do about it.

Reply:
> This is the way it's implemented for nats and rabbitmq also. I'll simplify this, and also make the change in those two files as well.

Let's not use this PR for fixing other tests. You can file a separate PR for that.